File: src/57-insert-interval/solution.rb
# @param {Integer[][]} intervals
# @param {Integer[]} new_interval
# @return {Integer[][]}
def insert(intervals, new_interval)
  res = []
  i, n = 0, intervals.length
  while i < n
    if new_interval.last < intervals[i].first
      res << new_interval
      return res + intervals[i..n-1]
    elsif new_interval.first > intervals[i].last
      res << intervals[i]
    else
      new_interval = [
        [intervals[i].first, new_interval.first].min,
        [intervals[i].last, new_interval.last].max
      ]
    end

    i += 1
  end
  res << new_interval
  res
end

File: src/57-insert-interval/solution.ts
function insert(intervals: number[][], newInterval: number[]): number[][] {
  let res: number[][] = [];
  let i: number = 0, n: number = intervals.length;

  while(i < n && newInterval[0] > intervals[i][1]) {
    res.push(intervals[i]);
    i++;
  }

  while(i < n && intervals[i][0] <= newInterval[1]) {
    newInterval = [
      Math.min(intervals[i][0], newInterval[0]),
      Math.max(intervals[i][1], newInterval[1]),
    ];
    i++;
  }

  res.push(newInterval);

  while(i < n) {
    res.push(intervals[i]);
    i++;
  }

  return res;
};
